### ISSUE TITLE
mismatch Content-Length: fetching something without size

### ISSUE THREAD

**Issue by Skarlso** [oras-project/oras-go#822]
Hello!

How do I use oras if I don't know the size of the object I'm trying to fetch?

This is the function:
```go
Fetch(ctx context.Context, desc ocispec.Descriptor) (io.ReadCloser, error)
```

Calling this with a descriptor like this:
```go
	desc := artdesc.Descriptor{
		MediaType: mimeType,
		Digest:    digest,
	}
```

The error I'm getting:

```
failed to fetch descriptor: GET "https://ghcr.io/v2/skarlso/ocm-podify/component-descriptors/github.com/weaveworks/podify/manifests/sha256:d5c11de4b76d11d786ead7507b4018cf3e7d7a2812c39eaf5262e8a479d81119": mismatch Content-Length
```

The offending code:

```go
	if size := resp.ContentLength; size != -1 && size != target.Size {
		
		fmt.Println("THE ERROR IS HERE")
		return nil, fmt.Errorf("%s %q: mismatch Content-Length", resp.Request.Method, resp.Request.URL)
	}
```

But I don't know the size of the object before fetching. How would I use this?

Thanks!

**Comment by Skarlso** [comment]
I think this might actually be a bug. I believe that `-1` check should be on the Target instead of the size.

```go
	if size := resp.ContentLength; target.Size != -1 && size != target.Size {
		
		fmt.Println("THE ERROR IS HERE")
		return nil, fmt.Errorf("%s %q: mismatch Content-Length", resp.Request.Method, resp.Request.URL)
	}
```

Which would effectively ignore that target.Size if it's set to `-1`. Which is UNKNOWN.

**Comment by Wwwsylvia** [comment]
Hi @Skarlso, thanks for opening the issue! The size check is by design and is a security consideration.

If you know the digest of a manifest but don’t have the descriptor, you have two options to fetch a manifest:

1. [`Repository.Resolve`](https://pkg.go.dev/oras.land/oras-go/v2@v2.5.0/registry/remote#Repository.FetchReference) first, then [`Repository.Fetch`](https://pkg.go.dev/oras.land/oras-go/v2@v2.5.0/registry/remote#Repository.Fetch)
`Repository.Resolve` resolves the tag/digest to a descriptor that can be passed to `Repository.Fetch`. 
However, based on the comment in the referencd issue, it seems this might not be your preferred option.

2. Call [`Repository.FetchReference`](https://pkg.go.dev/oras.land/oras-go/v2@v2.5.0/registry/remote#Repository.FetchReference)\
`Repository.FetchReference` allows fetching by tag/digest directly without needing the descriptor.
Could you try this and see if it helps in your scenario? Additionally, you may take a look at [`oras.FetchBytes`](https://pkg.go.dev/oras.land/oras-go/v2@v2.5.0#FetchBytes) if needed, which is a convenient utility function that wrap `FetchReference`.

**Comment by Skarlso** [comment]
Ah, I thought that might be it, and I designed my thing to do a Resolve first.

I guess that's fair. :) So Resolve the digest first, then do a Fetch with the resolved Descriptor is that right?

**Comment by Wwwsylvia** [comment]
> So Resolve the digest first, then do a Fetch with the resolved Descriptor is that right?

Yes, here is an [example](https://pkg.go.dev/oras.land/oras-go/v2@v2.5.0/registry/remote#example-Repository.Fetch-ManifestByDigest) that demonstrates this use case.

Something like
```go
exampleDigest := "sha256:b53dc03a49f383ba230d8ac2b78a9c4aec132e4a9f36cc96524df98163202cc7"

descriptor, err := repo.Resolve(ctx, exampleDigest)
if err != nil {
	panic(err)
}
rc, err := repo.Fetch(ctx, descriptor)
if err != nil {
	panic(err)
}
```

**Comment by Skarlso** [comment]
Ah thanks. Interesting, if my repo is already primed with the right ref, it's enough to give the resolve the digest and I don't have to provide the whole ref again? I didn't know that. That actually makes things a lot simpler. :) I don't have to pass around the ref.

**Comment by Wwwsylvia** [comment]
> if my repo is already primed with the right ref, it's enough to give the resolve the digest and I don't have to provide the whole ref again

This is correct. `Resolve` can take either a full reference like `myregistry.com/myrepo@sha256:abc123` or just a digest/tag like `sha256:abc123`.

**Comment by Skarlso** [comment]
Nice! Thanks, that helps a lot in keeping my code clean.

Thank you for the help! :)
